platform/cxxsupport/TOOLCHAIN_ARMC5/tuple

```diff
@@ -0,0 +1,777 @@
+                                           conjunction<negation<we_are_convertible_from<const tuple<U> &...>>,
+                                                       negation<conjunction<is_constructible<T, const tuple<U> &>...>>,
+                                                       negation<conjunction<is_same<T, U>...>>
+                                                      >
+                                      >
+                          >::value, bool> = true>
+    explicit tuple_base(const tuple<U...> &other) : tuple_elem<I, T>(get<I>(other))... { }
+    /* Converting move constructor */
+    template <typename... U,
+        enable_if_t<
+               conjunction<bool_constant<sizeof...(T) == sizeof...(U)>,
+                           is_constructible<T,const U&&>...,
+                           disjunction<bool_constant<sizeof...(T) != 1>,
+                                           conjunction<negation<we_are_convertible_from<tuple<U>...>>,
+                                                       negation<conjunction<is_constructible<T, tuple<U>>...>>,
+                                                       negation<conjunction<is_same<T, U>...>>
+                                                      >
+                                      >
+                          >::value, bool> = true>
+    explicit tuple_base(tuple<U...> &&other) : tuple_elem<I, T>(std::forward<U>(get<I>(other)))... { }
+    /* Converting pair copy constructor */
+    template <typename U1, typename U2,
+        enable_if_t<
+               conjunction<bool_constant<sizeof...(T) == 2>,
+                           we_are_constructible_from<const U1 &, const U2 &>
+                          >::value, bool> = true>
+    explicit tuple_base(const pair<U1,U2> &pair) :
+        tuple_base(pair.first, pair.second) { }
+    /* Converting pair move constructor */
+    template <typename U1, typename U2,
+        enable_if_t<
+               conjunction<bool_constant<sizeof...(T) == 2>,
+                           we_are_constructible_from<U1 &&, U2 &&>
+                          >::value, bool> = true>
+    explicit tuple_base(pair<U1,U2> &&pair) :
+        tuple_elem<0, tuple_element_t<0, tuple<T...>>>(std::forward<U1>(pair.first)),
+        tuple_elem<1, tuple_element_t<1, tuple<T...>>>(std::forward<U2>(pair.second)) { }
+    /* Copy constructor */
+    tuple_base(const tuple_base &other) = default;
+    /* Move constructor */
+    //tuple_base(tuple_base &&other) = default; // XXX ARMC5 does not support
+    tuple_base(tuple_base &&other) : tuple_elem<I,T>(std::forward<tuple_elem<I,T>>(other))... { }
+
+    void ignore(int...) { }
+
+    tuple_base &operator=(const tuple_base &other) = default;
+
+    //tuple_base &operator=(tuple_base &&other) = default; // XXX ARMC5 does not support
+    tuple_base &operator=(tuple_base &&other) {
+        ignore((((void)(*static_cast<tuple_elem<I, T> *>(this) = std::forward<tuple_elem<I,T>>(other))),1)...);
```

Comment:
Aha, brilliant!

Reply:
Note that if tuple is composed of pod types an optimisation might be to use memcpy. I'm not sure the compiler is able to do that by itself.

Reply:
Good point - it may be worth doing a few key optimisations like that. I very much doubt it would figure out it was a memcpy itself from that code.

To some extent, this was just "make it compile", in the knowledge that if we start using any of this, ARMC5 will start losing ground against the other compilers. But certainly worth special-casing that at least.

Reply:
Actually, if it is trivially copyable, we can just omit the explicit move operator, and let it use its default copy.